test/e2e/token_inactivity_timeout_test.go

```diff
@@ -0,0 +1,233 @@
+	oauthClientName := "demooauthclient"
+	redirectURIs := []string{"https://localhost"}
+	identityName := idpName + ":" + userName
+
+	uid, cleanup := createUser(t, userClient, userName, identityName)
+	defer cleanup()
+
+	cleanup = createIdentity(t, userClient, userName, identityName, idpName, uid)
+	defer cleanup()
+
+	cleanup = createOAuthClient(t, oauthClientClient, oauthClientName, redirectURIs)
+	defer cleanup()
+
+	shortLivedOAuthAccessToken := &oauthapi.OAuthAccessToken{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "Wf0AA1DphiD1FFLQV7BQIsesaaba-15RoSPZ3SvvXXS",
+		},
+		ClientName:               oauthClientName,
+		ExpiresIn:                defaultAccessTokenMaxAgeSeconds,
+		Scopes:                   []string{"user:full"},
+		RedirectURI:              redirectURIs[0],
+		UserName:                 userName,
+		UserUID:                  string(uid),
+		AuthorizeToken:           "mJOQ7Es5l9V7WYDl0bvl3E_hRjnJ21ZZxXH6YZj3yeS",
+		InactivityTimeoutSeconds: 300,
+	}
+
+	longLivedAccessToken := &oauthapi.OAuthAccessToken{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "Wf0AA1DphiD1FFLQV7BQIsesaaba-15RoSPZ3SvvXXT",
+		},
+		ClientName:               oauthClientName,
+		ExpiresIn:                defaultAccessTokenMaxAgeSeconds,
+		Scopes:                   []string{"user:full"},
+		RedirectURI:              redirectURIs[0],
+		UserName:                 userName,
+		UserUID:                  string(uid),
+		AuthorizeToken:           "mJOQ7Es5l9V7WYDl0bvl3E_hRjnJ21ZZxXH6YZj3yeT",
+		InactivityTimeoutSeconds: 300,
+	}
+
+	for _, accessToken := range []*oauthapi.OAuthAccessToken{shortLivedOAuthAccessToken, longLivedAccessToken} {
+		_, err = oauthClientClient.OAuthAccessTokens().Create(context.Background(), accessToken, metav1.CreateOptions{})
+		require.NoError(t, err)
+		defer func(name string) {
+			if err := oauthClientClient.OAuthAccessTokens().Delete(context.Background(), name, metav1.DeleteOptions{}); err != nil {
+				t.Logf("%v", err)
+			}
+		}(accessToken.Name)
+	}
+
+	httpTransport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		TLSHandshakeTimeout: 10 * time.Second,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+
+	req, err := http.NewRequest(http.MethodGet, kubeConfig.Host+"/apis/user.openshift.io/v1/users/~", &bytes.Buffer{})
+	require.NoError(t, err)
+	req.Header.Set("Authorization", "Bearer "+shortLivedOAuthAccessToken.Name)
+
+	resp, err := httpTransport.RoundTrip(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("short lived access token should work. Recieved status %v", resp.Status)
+	}
+
+	// Wait for token to expire.
+	fmt.Println("Waiting for short lived access token to expire due to inactivity")
+	// Nanosecond delta make sure that test does not hit edge cases.
+	time.Sleep(300*time.Second + time.Nanosecond)
```

Comment:
This `time.Nanosecond` won't make the difference. How often does the timeout controller run within kube-apiserver? Would add at least one iteration duration here.

Reply:
The timeout controller runs every 100s. But if a token is past its validity, it expires the token instantaneously rather than waiting for next interval. https://github.com/openshift/kubernetes/blob/4d82d07bf0c5372c069fbc5dd60baa9eed3fac09/openshift-kube-apiserver/authentication/oauth/timeoutvalidator.go#L218-L230